ingestor/adx/tasks.go

```diff
@@ -288,14 +294,274 @@ func (t *SummaryRuleTask) Run(ctx context.Context) error {
+			cnd.Reason = "Succeeded"
+		case "Failed":
+			// If the operation failed, we need to check if we should retry
+			if status.ShouldRetry {
+				// Kuso thinks it's worth retrying. We'll set the management submission
+				// condition to False, which will have the effect of re-submitting the command.
+				cmdCnd := rule.GetCondition()
+				if cmdCnd != nil {
+					cmdCnd.Status = metav1.ConditionFalse
+					cmdCnd.Reason = "Failed"
+					cmdCnd.Message = "Async operation failed, but should be retried"
+					cmdCnd.LastTransitionTime = metav1.Time{Time: time.Now()}
+					rule.SetCondition(*cmdCnd)
+				}
+
+				// We're going to resubmit this task, so we can keep the async
+				// operation's status as Unknown, though we'll update the Reason.
+				cnd.Reason = "Retry"
+			} else {
+				// This is a terminal failure, so we'll update the condition accordingly.
+				cnd.Status = metav1.ConditionFalse
+				cnd.Reason = "Failed"
+			}
+
+		case "Scheduled", "InProgress":
+			// Just record what's happening and we'll check again later.
+			cnd.Reason = status.State
+
+		case "Throttled":
+			// NOTE:
+			// We're being throttled, so this async operation will not be retried
+			// and we shouldn't retry either. We need to log this and set the status.
+			// This should be an actionable state. Triage should occur to determine
+			// if we're doing something too aggressively or if we need to scale out
+			// to accommodate the load.
+			logger.Errorf("Kusto operations are being throttled: %s", t.kustoCli.Database())
+			cnd.Reason = status.State
+			cnd.Status = metav1.ConditionFalse
+
+		default:
+			// All other reasons are non-actionable. We simply record what has occured
+			// and move on. In all cases, the async operation is considered to be failed
+			// since it's in a non-actionable non-succeeding state.
+			cnd.Reason = status.State
+			cnd.Status = metav1.ConditionFalse
+		}
+
+		rule.SetOperationIDCondition(*cnd)
+		if err = t.store.UpdateStatus(ctx, &rule, err); err != nil {
+			return fmt.Errorf("failed to update summary rule status: %w", err)
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+func (t *SummaryRuleTask) shouldRetryTask(ctx context.Context, operationId, state string) (bool, error) {
+	// For reasons I don't understand, the GO SDK is not able to execute the query
+	// `.show operations | where OperationId == "operationId" | summarize arg_max(LastUpdatedOn, State, ShouldRetry)`.
+	// The client library itself throws an error stating that ShouldRetry should be a bool but is a float64.
+	// Even if I cast ShouldRetry to a bool, the error persists.
+	//
+	// This method is a work-around where we query if the operation should be retried.
+	// `.show operations | where OperationId == "operationId" | where State == "state" | where ShouldRetry == true | count`
+	// If the value is non zero, we should retry.
+	stmt := kql.New(".show operations").
+		AddLiteral(" | where OperationId == ").AddString(operationId).
```

Comment:
This should be bulk oriented vs per single operation so it doesn't spawn tons of queries if many need to be retried.